File: 01-basic-auth/node/src/repository/memoryUserRepository.ts
import { User } from '../domain/user';
import { UserRepository } from '../domain/repository';

export class MemoryUserRepository implements UserRepository {
  private users = new Map<string, User>();

  async create(user: User): Promise<void> {
    this.users.set(user.id, user);
  }

  async getById(id: string): Promise<User | null> {
    return this.users.get(id) ?? null;
  }

  async getByEmail(email: string): Promise<User | null> {
    for (const user of this.users.values()) {
      if (user.email === email) return user;
    }
    return null;
  }

  async delete(id: string): Promise<void> {
    this.users.delete(id);
  }
}

File: 01-basic-auth/csharp/Repository/MemoryUserRepository.cs
using BasicAuth.Domain;

namespace BasicAuth.Repository;

// In-memory реализация IUserRepository.
//
// Важно:
// - данные живут только в памяти процесса (после перезапуска исчезнут)
// - GetByEmail использует перебор (O(n)) — для учебного примера достаточно
public class MemoryUserRepository : IUserRepository
{
    private readonly Dictionary<string, User> _users = new();

    public Task CreateAsync(User user)
    {
        _users[user.Id] = user;
        return Task.CompletedTask;
    }

    public Task<User?> GetByIdAsync(string id)
    {
        _users.TryGetValue(id, out var user);
        return Task.FromResult(user);
    }

    public Task<User?> GetByEmailAsync(string email)
    {
        var user = _users.Values.FirstOrDefault(u => u.Email == email);
        return Task.FromResult(user);
    }

    public Task DeleteAsync(string id)
    {
        _users.Remove(id);
        return Task.CompletedTask;
    }
}
